File: impl-large/python/domain.py
"""Task Service, large tier — domain types, constants and pure rules."""

from dataclasses import dataclass

MAX_TITLE_LENGTH = 80
MAX_NAME_LENGTH = 60
MAX_COMMENT_LENGTH = 200
MAX_BULK_ITEMS = 20
MIN_PRIORITY = 1
MAX_PRIORITY = 5
DEFAULT_LIMIT = 20
MAX_LIMIT = 100
DEFAULT_QUOTA = 10000
PROBE_QUOTA = 5
PORT = 8080

ROLES = ("admin", "user")
STATUSES = ("todo", "in_progress", "done", "archived")
STATUS_BONUS = {"todo": 0, "in_progress": 3, "done": 5, "archived": 0}
TRANSITIONS = {
    ("todo", "in_progress"), ("todo", "archived"), ("in_progress", "todo"),
    ("in_progress", "done"), ("done", "archived"),
}
ACTIONS = ("create", "update", "delete", "restore")
PROJECT_SORTS = ("id", "name", "taskCount")
TASK_SORTS = ("id", "title", "priority", "score", "status")
USER_SORTS = ("id", "username", "role")
COMMENT_SORTS = ("id", "authorId")
SEQ_SORTS = ("seq",)
GROUP_BYS = ("assignee", "status", "project")


@dataclass
class User:
    id: int
    username: str
    password: str
    role: str
    quota: int
    version: int = 1
    deleted: bool = False


@dataclass
class Session:
    token: str
    userId: int
    used: int = 0


@dataclass
class Project:
    id: int
    name: str
    ownerId: int
    version: int = 1
    deleted: bool = False


@dataclass
class Task:
    id: int
    projectId: int
    title: str
    priority: int
    status: str
    assigneeId: int | None
    internalNote: str = ""
    version: int = 1
    deleted: bool = False


@dataclass
class Comment:
    id: int
    taskId: int
    authorId: int
    body: str


@dataclass
class AuditEntry:
    seq: int
    actorId: int
    action: str
    resource: str
    resourceId: int


@dataclass
class OutboxEvent:
    seq: int
    name: str
    resourceId: int
    delivered: bool = False


class AppError(Exception):
    """Every failure path raises this. The api layer turns it into the envelope."""

    def __init__(self, status: int, code: str, message: str, details: list | None = None):
        self.status = status
        self.code = code
        self.message = message
        self.details = details or []


def bad_request() -> AppError:
    return AppError(400, "bad_request", "the request is malformed")


def unauthorized() -> AppError:
    return AppError(401, "unauthorized", "authentication is required")


def invalid_credentials() -> AppError:
    return AppError(401, "invalid_credentials", "the username or password is wrong")


def forbidden() -> AppError:
    return AppError(403, "forbidden", "you may not access this resource")


def not_found() -> AppError:
    return AppError(404, "not_found", "the resource does not exist")


def conflict() -> AppError:
    return AppError(409, "conflict", "the resource already exists")


def invalid_transition() -> AppError:
    return AppError(409, "invalid_transition", "the status change is not allowed")


def precondition_failed() -> AppError:
    return AppError(412, "precondition_failed", "the resource has changed")


def precondition_required() -> AppError:
    return AppError(428, "precondition_required", "the If-Match header is required")


def quota_exceeded() -> AppError:
    return AppError(429, "quota_exceeded", "the request quota is exhausted")


def invalid(details: list) -> AppError:
    details.sort(key=lambda entry: (entry["field"], entry["message"]))
    return AppError(422, "validation_failed", "the request body is not valid", details)


def fail(field: str, message: str) -> dict:
    return {"field": field, "message": message}


def compute_score(priority: int, status: str) -> int:
    base_score = priority * 10
    return base_score + STATUS_BONUS[status]


def check_string(value: str, field_name: str, max_length: int, errors: list) -> None:
    if value == "":
        errors.append(fail(field_name, f"{field_name} is required"))
    elif len(value) > max_length:
        errors.append(fail(field_name, f"{field_name} is too long"))


def check_priority(value: int | None, errors: list) -> None:
    if value is None or value < MIN_PRIORITY or value > MAX_PRIORITY:
        errors.append(fail("priority", "priority is out of range"))


def check_status(value, errors: list) -> None:
    if not isinstance(value, str) or value not in STATUSES:
        errors.append(fail("status", "status is not valid"))


def check_role(value, errors: list) -> None:
    if not isinstance(value, str) or value not in ROLES:
        errors.append(fail("role", "role is not valid"))


def check_quota(value, errors: list) -> None:
    if not isinstance(value, int) or isinstance(value, bool) or value < 0:
        errors.append(fail("quota", "quota is out of range"))

File: impl-large/ruby/domain.rb
# frozen_string_literal: true

# Task Service, large tier — domain types, constants and pure rules.

module Domain
  MAX_TITLE_LENGTH = 80
  MAX_NAME_LENGTH = 60
  MAX_COMMENT_LENGTH = 200
  MAX_BULK_ITEMS = 20
  MIN_PRIORITY = 1
  MAX_PRIORITY = 5
  DEFAULT_LIMIT = 20
  MAX_LIMIT = 100
  DEFAULT_QUOTA = 10_000
  PROBE_QUOTA = 5
  PORT = 8080

  ROLES = %w[admin user].freeze
  STATUSES = %w[todo in_progress done archived].freeze
  STATUS_BONUS = { "todo" => 0, "in_progress" => 3, "done" => 5, "archived" => 0 }.freeze
  TRANSITIONS = [%w[todo in_progress], %w[todo archived], %w[in_progress todo],
                 %w[in_progress done], %w[done archived]].freeze
  ACTIONS = %w[create update delete restore].freeze
  PROJECT_SORTS = %w[id name taskCount].freeze
  TASK_SORTS = %w[id title priority score status].freeze
  USER_SORTS = %w[id username role].freeze
  COMMENT_SORTS = %w[id authorId].freeze
  SEQ_SORTS = %w[seq].freeze
  GROUP_BYS = %w[assignee status project].freeze

  User = Struct.new(:id, :username, :password, :role, :quota, :version, :deleted)
  Session = Struct.new(:token, :user_id, :used)
  Project = Struct.new(:id, :name, :owner_id, :version, :deleted)
  Task = Struct.new(:id, :project_id, :title, :priority, :status, :assignee_id,
                    :internal_note, :version, :deleted)
  Comment = Struct.new(:id, :task_id, :author_id, :body)
  AuditEntry = Struct.new(:seq, :actor_id, :action, :resource, :resource_id)
  OutboxEvent = Struct.new(:seq, :name, :resource_id, :delivered)

  # Every failure path raises this. The api layer turns it into the envelope.
  class AppError < StandardError
    attr_reader :status, :code, :details

    def initialize(status, code, message, details = [])
      super(message)
      @status = status
      @code = code
      @details = details
    end
  end

  module_function

  def bad_request
    AppError.new(400, "bad_request", "the request is malformed")
  end

  def unauthorized
    AppError.new(401, "unauthorized", "authentication is required")
  end

  def invalid_credentials
    AppError.new(401, "invalid_credentials", "the username or password is wrong")
  end

  def forbidden
    AppError.new(403, "forbidden", "you may not access this resource")
  end

  def not_found
    AppError.new(404, "not_found", "the resource does not exist")
  end

  def conflict
    AppError.new(409, "conflict", "the resource already exists")
  end

  def invalid_transition
    AppError.new(409, "invalid_transition", "the status change is not allowed")
  end

  def precondition_failed
    AppError.new(412, "precondition_failed", "the resource has changed")
  end

  def precondition_required
    AppError.new(428, "precondition_required", "the If-Match header is required")
  end

  def quota_exceeded
    AppError.new(429, "quota_exceeded", "the request quota is exhausted")
  end

  def invalid(details)
    details.sort_by! { |entry| [entry[:field], entry[:message]] }
    AppError.new(422, "validation_failed", "the request body is not valid", details)
  end

  # Named fail_with, not fail, because Kernel#fail is a Ruby built-in.
  def fail_with(field, message)
    { field: field, message: message }
  end

  def compute_score(priority, status)
    base_score = priority * 10
    base_score + STATUS_BONUS[status]
  end

  def check_string(value, field_name, max_length, errors)
    if value.empty?
      errors << fail_with(field_name, "#{field_name} is required")
    elsif value.length > max_length
      errors << fail_with(field_name, "#{field_name} is too long")
    end
  end

  def check_priority(value, errors)
    return if value.is_a?(Integer) && value >= MIN_PRIORITY && value <= MAX_PRIORITY

    errors << fail_with("priority", "priority is out of range")
  end

  def check_status(value, errors)
    errors << fail_with("status", "status is not valid") unless STATUSES.include?(value)
  end

  def check_role(value, errors)
    errors << fail_with("role", "role is not valid") unless ROLES.include?(value)
  end

  def check_quota(value, errors)
    return if value.is_a?(Integer) && !value.negative?

    errors << fail_with("quota", "quota is out of range")
  end
end
